### PR TITLE
Round for int type

### PR DESCRIPTION
**Pull request by varocarbas**
New` int.Round` + some int `Math.Round` overloads.

Example:` int.Round(123456, 4, MidpointRounding.AwayFromZero)` -> 123500

### CODE REVIEW

**Comment by varocarbas** [comment]
I did create an issue about this: https://github.com/dotnet/coreclr/issues/3204

**Comment by varocarbas** [comment]
I realised now that I also included here the modifications from my previous PR. It was a mistake (not too experienced with GitHub)

**Comment by varocarbas** [comment]
I am not completely sure, but it seems that the following situation isn't being accounted properly in the tests:

```
if (i == 0) return 0;
if ((digits <= 0) || (digits > maxRoundingDigitsInt)) return 0;
```

These are two independent conditions, one after the other (perfectly understood when compiling mscorlib in Windows). But apparently, the online tests misunderstand that the second condition is included in the first one and, consequently, the following correction is required:

```
if (i == 0) { return 0;}
if ((digits <= 0) || (digits > maxRoundingDigitsInt)) return 0;
```

**Comment by varocarbas** [comment]
No idea what is the problem with the Ubuntu server, but it keeps delivering errors for no reason. The last time, it had failed with exactly the same version which suceeded just 1 day ago. Additionally, the details say that there is no identified problems. Apparently there is an error with this server ("No test result found" = it cannot send the results back?!).

In any case, some of my last modifications (involving 1-line conditions) seemed to have triggered errors in other servers too; perhaps you might want to take a look at it.

**Comment by linuxuser27** [comment]
I am unclear what this example means:
int.Round(123456, 4, MidpointRounding.AwayFromZero) -> 1235

Rounding is pretty well defined with respect to integers and this example is not rounding but rather truncation. It is possible you meant 123500, but an answer of 1235 is definitely not rounding as I have ever observed it.

Can you clarify what problem this version of Round is attempting to address?

**Comment by varocarbas** [comment]
@linuxuser27 Actually you are right. `int.Round` is too confusing.
On the other hand, this isn't truncation either. Truncation would be 1234.

The whole point of this implementation is extending the current decimal-rounding .NET capabilities to non-decimal numbers. Although the functionality is virtually identical (other than by the decimal point), the name isn't.

On the other hand, including a proper integer rounding (as you suggest) might also be worthy (and the proposed algorithm can be used). Bear in mind that this is the first PR of an issue proposing changes in various `Math.System` functions.

In any case, the whole point of this proposal was practically extending existing functionalities, which don't have to necessarily agree with a mathematical (or other) concept. For example, what `Math.Truncate` delivers is not exactly truncation (but truncation of decimal digits in decimal numbers).

**Comment by varocarbas** [comment]
@linuxuser27 PS: bear in mind that this PR doesn't just include `int.Round`, but also new overloads in `Math.Round` for `int` arguments. In fact, these overloads were precisely the main reason for this PR; I included the `int.Round` part, just to keep it consistent with other types. Also I referred `int.Round` in the title because is shorter (than various overloads of...).

**Comment by linuxuser27** [comment]
> Why .NET makes a diference? The functionality doesn't need to necessarily agree with a mathematical (or other) concept; although you are right and the name is too confusing.

@varocarbas I somewhat disagree with this statement. You are adding to the Math library so I would argue that it absolutely must adhere to the concept - if the function is new and there is no compelling reason to not adhere to it. Existing functionality may not, but we should endeavor to be rigorous when possible. 

> ... including a proper integer rounding (as you suggest) might also be worthy (and the proposed algorithm might be used too). 

'Might' doesn't seem like a good enough reason to break the fundamental concept of what rounding means. Your implement for int.Round() isn't rounding and isn't intuitive, it is actually removing orders of magnitude for an ordinal value. If this was for a string, then I get it, but for an integer this seems wrong.

**Comment by linuxuser27** [comment]
Do be clear about my intentions here. I am not saying this is a terrible idea, I am merely trying to understand the core issue and if the runtime needs this kind of functionality. The rest of the community needs to weight in.

**Comment by varocarbas** [comment]
@linuxuser27  As explained in my second comment, the primary implementation is not `int.Round`, but new overloads of `Math.Round` by exactly emulating what is being done for other types (= quite intuitive to me).

> If this was for a string, then I get it, but for an integer this seems wrong. -> Actually, I was always thinking of strings. Truncation (and this is not exactly truncation, but kind of) is precisely meant to deal with string. Better: string-like situations, where you can remove/add digits. Note that this kind of numeric removal is much quicker than using `Substring`.

In any case, you are making a quite valid point. Honestly, I just wanted to speed up a bit the associated issue by showing some code; and perhaps I didn't think carefully about what functions in `System.Math `are expected to deliver. On the other hand, there are some methods like the referred `Math.Truncate()` which aren't precisely an accurate implementation of what is expected (i.e., why converting 1.2345 into 1 rather than into 1.234, as opposed to the rounded 1.235 version?).

**Comment by varocarbas** [comment]
@linuxuser27 The whole point of this (and the other proposals in the associated issue) is removing the type constraints of quite a few `System.Math` functions, such that they can also be used with integer types. I have seen myself in lots of situations where I am forced to perform 2 type conversions to use a function. For example: I am dealing with `long` values and want to raise 123456 to the 1/9th power, I would have to convert the variable to `double`, use `Math.Pow` and then back to `long`. The same thing happens with any trigonometrical function or with other methods (or with the referred `Math.Truncate` or `Math.Round`).

Some of these limitations are more or less logical on account of the underlying mathematical ideas (e.g., your pretty good point), but some of them aren't. And in any case, I see it as an unnecessary type conversion which might be avoided in quite a few scenarios, on account of the basic idea that you can use integer values in almost any situation where you can use decimal ones. So, why decimal numbers have so much advantage in `System.Math`?

**Comment by linuxuser27** [comment]
Okay. I think having an overload for `Math.Pow()` that takes and returns an integral type is useful and has value. That being said, I don't see a lot of value with implementing `Math.Round()` on an integral type, but that is just me. My suggestion would be to start with `Math.Pow()` and other functions that have better defined semantics.

**Comment by varocarbas** [comment]
@linuxuser27 As said, you made a fair point. That's why I modified the code (+ the title and description) to use `int.Round` and `Math.Round` to deliver proper integer rounding. 

I have also written my thoughts about how to account for the proposed truncation (also for decimal types), together with other modifications of `System.Math `methods in the last issue associated with this PR. You might want to take a look there: https://github.com/dotnet/corefx/issues/6308#issuecomment-187654590

**Comment by krytarowski** [comment]
Is this a standard operation available anywhere else? If not, why is it needed here?

**Comment by varocarbas** [comment]
@krytarowski Note that this is just one of the proposed modifications. You can find an updated list and quite a few comments in the new CoreFX issue (https://github.com/dotnet/corefx/issues/6308).

Regarding your question of being available anywhere else, I understand that you mean other programming languages. It should be the case, but honestly no idea and don't think that this fact is relevant at all. This is a valid mathematical operation (integer rounding) which complements the incomplete `Math.Round` method (only accounting for decimal rounding), what is more than enough to justify its inclusion: making the .NET rounding approach more comprehensive and fully compatible with the actual mathematical rounding.

On the other hand, I didn't think about any of the aforementioned points before proposing this modification (or the other ones in the aforementioned issue). I exclusively focused on merely practical and useful as per my programming expertise functionality. Now, if you want to round (or truncate -> also proposed a modification there) in .NET you would have to use a string-based approach in quite a few scenarios; to not mention the fact that you would have to implement the given functionality yourself. For example: you are dealing with many integers and you want to account only for the first 6 digits, the only option you have now is converting to string, use `Substring` (or similar) and convert back to numeric format. Alternatively, you might implement a numeric-based approach (i.e., like the one I am proposing here), although this is less straightforward. 

In summary, this is a reasonably-common requirement (-> useful), which additionally represents a valid mathematical methodology (-> comprehensive) and .NET is currently accounting for it only partially via decimal rounding (-> consistent).

**Comment by krytarowski** [comment]
> Regarding your question of being available anywhere else, I understand that you mean other programming languages. It should be the case, but honestly no idea and don't think that this fact is relevant at all. 

I disagree. rounding is well defined in other programing languages usually since their inception it means the same. For example in C:

```
$ man round
ROUND(3)                   Library Functions Manual                   ROUND(3)

NAME
     round, roundf, roundl - round to nearest integral value

LIBRARY
     Math Library (libm, -lm)

SYNOPSIS
     #include <math.h>

     double
     round(double x);

     float
     roundf(float x);

     long double
     roundl(long double x);

DESCRIPTION
     The round(), roundf(), and roundl() functions return the nearest integral
     value to x; if x lies halfway between two integral values, then these
     functions return the integral value with the larger absolute value (i.e.,
     they round away from zero).

SEE ALSO
     ceil(3), floor(3), math(3), rint(3), trunc(3)

STANDARDS
     The round() and roundf() functions conform to ISO/IEC 9899:1999
     (``ISO C99'').

HISTORY
     The round() and roundf() functions appeared in NetBSD 2.0.

NetBSD 7.99                    November 12, 2013                   NetBSD 7.99
```

**Comment by krytarowski** [comment]
I asked about other languages because this operation looks very customized and isn't needed to be defined by a library. You can implement your own version of rounding() the way you like pretty in 1-liner.

`Example:int.Round(123456, 4, MidpointRounding.AwayFromZero) -> 123500`

```
$ cat test.c                                                                                                                                          
#include <math.h>
#include <stdio.h>

int main(int argc, char **argv)
{
	float f = roundf(123456.0f / powf(10.0, 2.0)) * powf(10.0, 2.0);
	printf("%f\n", f);
}
$ gcc test.c -o test 
$ ./test             
123500.000000
```

and problem solved.

**Comment by krytarowski** [comment]
It has even similar number of characters to `int.Round(123456, 4, MidpointRounding.AwayFromZero)`

**Comment by varocarbas** [comment]
@krytarowski Emulating what other languages might be helpful to get idea of the kind of functionalities which are more likely to be useful. On the other hand, there are many functionalities in many programming languages which are virtually never used and might even be removed at further points. That's why, this information should better be completed with some stats about the usage of the given methods.

But the previous paragraph seems to make only sense at the start, when you are firstly thinking about creating a programming language; mainly if you don't have a relevant programming expertise. If the language is mature enough and you are knowledgeable enough why wondering about what other languages do?!

Anyway, one of the methods you refer (rint -> http://en.cppreference.com/w/cpp/numeric/math/rint) supports my implementation.

**Comment by krytarowski** [comment]
`rint`(3) is something different, but of course it may be used to define custom integer rounding.

**Comment by varocarbas** [comment]
Regarding your

>  pretty in 1-liner.
> 
> Example:int.Round(123456, 4, MidpointRounding.AwayFromZero) -> 123500
> 
> $ cat test.c                                                                                                                                          
> #include <math.h>
> #include <stdio.h>
> 
> int main(int argc, char **argv)
> {
>     float f = roundf(123456.0f / powf(10.0, 2.0)) * powf(10.0, 2.0);
>     printf("%f\n", f);
> }
> $ gcc test.c -o test 
> $ ./test             
> 123500.000000
> and problem solved.
> 

This code will most likely perform much worse than the method I am proposing (well... its .NET equivalence; you should better write .NET codes when participating in a discussion about .NET code): you are making multiple calls (including type conversions) to in-built functions, rather than one as I am requesting. To not mention the fact that you rely on floating point (and also the .NET equivalences will do) what might provoke rounding problems in quite a few situations.

On the other hand, if your approach (I can do it myself) would be applicable to anything, there wouldn't any in-built methods. For example: most of the `System.Math` functions would be redundant on account of these ideas (i.e., they perform simplistic extensions of other ones).

**Comment by krytarowski** [comment]
I disagree with this PR. But let see what others thing.

**Comment by varocarbas** [comment]
@krytarowski OK. I re-invite you to take a look at the associated CoreFX issue discussing about other `System.Math` methods.

**Comment by adityamandaleeka** [comment]
I'm closing this PR. Based on the latest commits (after the previous comment), it's no longer relevant and in any case, @varocarbas is working on an API proposal on the CoreFX repo.

**Comment by varocarbas** [comment]
@adityamandaleeka Sure, no problem. 
Just to clarify that varocarbas is me :) The CoreFX issue is the result of an evolution which was precisely started here.
